File: libKh/KhStream.cs
using System;
using System.IO;

namespace Kh
{
<<<<<<< HEAD
    public class FileStream : System.IO.Stream
    {
        IDX.FileEntry entry;

        long position;

        /// <summary>
        /// Open a Kingdom Hearts' file stream
        /// </summary>
        /// <param name="idx">IDX object</param>
        /// <param name="filename">name of the file to open</param>
        public FileStream(IDX idx, string filename)
        {
            position = 0;
            entry = idx.OpenFile(filename);
            if (entry.index == -1)
                throw new System.IO.FileNotFoundException();
=======
    class KhFileStream : System.IO.Stream
    {
        Stream internalStream;
        long internalPosition;

        long virtualPosition;
        long virtualLength;

        public KhFileStream(IDX idx, string filename)
        {
            internalPosition = 0;
            virtualPosition = 0;
>>>>>>> parent of 906e419... BAR Editor
        }

        public override long Position
        {
<<<<<<< HEAD
            get { return position; }
            set { if (CanSeek) position = value; }
        }
        public override long Length
        {
            get { return entry.length; }
=======
            get { return virtualPosition; }
            set { virtualPosition = value; }
        }
        public override long Length
        {
            get { return virtualLength; }
>>>>>>> parent of 906e419... BAR Editor
        }
        public override bool CanRead
        {
            get { return true; }
        }
        public override bool CanSeek
        {
<<<<<<< HEAD
            get { return entry.streamed == false; }
=======
            get { return true; }
>>>>>>> parent of 906e419... BAR Editor
        }
        public override bool CanWrite
        {
            get { return false; }
        }

        public override int Read(byte[] buffer, int offset, int count)
        {
<<<<<<< HEAD
            if (CanRead)
            {
                if (Position + count > Length)
                    count = (int)(Length - Position);
                entry.stream.Position = entry.position + Position;
                entry.stream.Read(buffer, offset, count);
                position += count;
                return count;
            }
=======
>>>>>>> parent of 906e419... BAR Editor
            return 0;
        }
        public override void Write(byte[] buffer, int offset, int count)
        {

        }
        public override void Flush()
        {

        }
        public override void SetLength(long value)
        {

        }

        public override long Seek(long offset, SeekOrigin origin)
        {
<<<<<<< HEAD
            if (CanSeek)
            {
                switch (origin)
                {
                    case SeekOrigin.Begin:
                        Position = offset;
                        break;
                    case SeekOrigin.Current:
                        Position += offset;
                        break;
                    case SeekOrigin.End:
                        Position = Length + offset;
                        break;
                }
=======
            switch (origin)
            {
                case SeekOrigin.Begin:
                    Position = offset;
                    break;
                case SeekOrigin.Current:
                    Position += offset;
                    break;
                case SeekOrigin.End:
                    Position = Length + offset;
                    break;
>>>>>>> parent of 906e419... BAR Editor
            }
            return Position;
        }
    }
}

File: libKhcpp/KhStream.cpp
#include "KhStream.h"


namespace Kh
{

	FileStream::FileStream(IDX *idx, const std::wstring &filename)
	{
		InitializeInstanceFields();
		position = 0;
		entry = idx->OpenFile(filename);
		if (entry.index == -1)
		{
			throw System::IO::FileNotFoundException();
		}
	}

	const long long &FileStream::getPosition() const
	{
		return position;
	}

	void FileStream::setPosition(const long long &value)
	{
		if (getCanSeek())
		{
			position = value;
		}
	}

	const long long &FileStream::getLength() const
	{
		return entry.length;
	}

	const bool &FileStream::getCanRead() const
	{
		return true;
	}

	const bool &FileStream::getCanSeek() const
	{
		return entry.streamed == false;
	}

	const bool &FileStream::getCanWrite() const
	{
		return false;
	}

	int FileStream::Read(unsigned char buffer[], int offset, int count)
	{
		if (getCanRead())
		{
			if (getPosition() + count > getLength())
			{
				count = static_cast<int>(getLength() - getPosition());
			}
			entry.stream->Position = entry.position + getPosition();
			entry.stream->Read(buffer, offset, count);
			position += count;
			return count;
		}
		return 0;
	}

	void FileStream::Write(unsigned char buffer[], int offset, int count)
	{

	}

	void FileStream::Flush()
	{

	}

	void FileStream::SetLength(long long value)
	{

	}

	long long FileStream::Seek(long long offset, SeekOrigin origin)
	{
		if (getCanSeek())
		{
			switch (origin)
			{
				case SeekOrigin::Begin:
					setPosition(offset);
					break;
				case SeekOrigin::Current:
					setPosition(getPosition() + offset);
					break;
				case SeekOrigin::End:
					setPosition(getLength() + offset);
					break;
			}
		}
		return getPosition();
	}

	void FileStream::InitializeInstanceFields()
	{
		position = 0;
	}
}
